stan/math/rev/mat/vectorize/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,153 @@
+#ifndef STAN_MATH_REV_MAT_VECTORIZE_APPLY_BINARY_SCALAR_HPP
+#define STAN_MATH_REV_MAT_VECTORIZE_APPLY_BINARY_SCALAR_HPP
+
+#include <stan/math/prim/mat/vectorize/apply_scalar_binary.hpp>
+#include <stan/math/rev/core/var.hpp>
+
+namespace stan {
+
+  namespace math {
+
+    /**
+     * Template specialization for vectorized functions applying to
+     * var and var arguments.
+     *
+     * @tparam F Type of function defining static apply function.
+     */
+    template <typename F>
+    struct apply_scalar_binary<F, stan::math::var, stan::math::var> {
+      /**
+       * Function return type, which is <code>var</code>.
+       */
+      typedef stan::math::var return_t;
+
+      /**
+       * Apply the function specified by F to the specified argument.
+       * This is defined through a direct application of
+       * <code>F::fun()</code>, which must be defined for var and var
+       * arguments.
+       *
+       * @param x var argument scalar.
+       * @param y var argument scalar.
+       * @return Result of applying F to the scalar x and y.
+       */
+      static inline return_t apply(const stan::math::var& x,
+      const stan::math::var& y) {
+        return F::fun(x, y);
+      }
+    };
+
+    /**
+     * Template specialization for vectorized functions applying to
+     * var and int arguments.
+     *
+     * @tparam F Type of function defining static apply function.
+     */
+    template <typename F>
+    struct apply_scalar_binary<F, stan::math::var, int> {
```

Comment:
This seems like way too many specializations.  Is there really no way to do this generically?  I'd have thought a single implementation for scalars and specializations for vectors would be enough.

Technically, we can push a lot of this down below the `mat` directory level into `arr` and `prim`, but I'm not going to worry about that now.

Reply:
I've been looking into this for the last few days. If I run the following script,
```
#include <Eigen/Dense>
#include <iostream>

template <typename T, int R, int C>
void test(Eigen::Matrix<T, R, C> m) {
  std::cout << "Hi" << std::endl;
}

template <typename T>
void test(T m) {
  std::cout << "bye" << std::endl;
}

int main(int argc, char **argv) {
  Eigen::MatrixXd a = Eigen::MatrixXd::Constant(5, 3, 2);
  Eigen::MatrixXd b = a.block(1, 1, 1, 1);
  test(a);
  test(a.block(1, 1, 1, 1));
  test(b);
  return(1);
}
``` 
using `clang++ -I . -isystem lib/eigen_3.3.3 -isystem lib/boost_1.62.0 -isystem lib/cvodes_2.9.0/include tmp.cpp && ./a.out`, I get
```
Hi
bye
Hi
```
Do I have to worry about the vectorized function being applied to `a.block` directly? If not, that will make things easier.

Reply:
If I have to worry about that case, what should the return type be? `Eigen::Block<Eigen::Matrix<T, R, C> >` or `Eigen::Matrix<T, R, C>`?

Reply:
I'm not sure what case you're talking about.  All of our return types now are `Eigen::Matrix<T, R, C>` because we haven't generalized our matrix function arguments to accept expression templates as arguments.

Reply:
Okay. I pushed the changes because I think it might be clearer what I'm referring to with the new code.